src/Interpreters/Cache/FileCache.cpp

```diff
@@ -136,7 +140,15 @@ const FileCache::UserInfo & FileCache::getInternalUser()
 
 bool FileCache::isInitialized() const
 {
-    return is_initialized.load(std::memory_order_seq_cst);
+    return is_initialized;
+}
+
+void FileCache::throwInitExceptionIfNeeded()
+{
+    if (!load_metadata_asynchronously && init_exception)
```

Comment:
First it is better to check `load_metadata_asynchronously` and return early.
Then use `std::lock_guard lock(init_mutex);` to avoid data races when other thread set `init_exception`.